sound/soc/sof/intel/hda-dai.c

```diff
@@ -475,17 +480,78 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
 			     struct snd_pcm_hw_params *params,
 			     struct snd_soc_dai *dai)
 {
+	struct ssp_dai_dma_data *dma_data;
+
+	dma_data = kzalloc(sizeof(*dma_data), GFP_KERNEL);
+	if (!dma_data)
+		return -ENOMEM;
+
+	snd_soc_dai_set_dma_data(dai, substream, dma_data);
+
 	return ssp_dai_setup_or_free(substream, dai, true);
 }
 
-static int ssp_dai_hw_free(struct snd_pcm_substream *substream,
+static int ssp_dai_prepare(struct snd_pcm_substream *substream,
 			   struct snd_soc_dai *dai)
 {
+	struct ssp_dai_dma_data *dma_data;
+	int ret = 0;
+
+	dma_data = snd_soc_dai_get_dma_data(dai, substream);
+	if (!dma_data) {
+		dev_err(dai->dev, "%s: failed to get dma_data\n", __func__);
+		return -EIO;
+	}
+
+	if (dma_data->suspended) {
+		dma_data->suspended = false;
+		ret = ssp_dai_setup_or_free(substream, dai, true);
```

Comment:
If ssp_dai_setup_or_free() failed then the flag should not have been changed to `false`

I think the `dai_ready` flag use would make the code a bit simpler

Reply:
Not sure if it matters. if the prepare fails, what happens? i

Reply:
done

Reply:
It is up to user space, but I would say that it is going to close the stream, but it might re-try with different parameters.